src/main/scala/com/databricks/spark/redshift/RedshiftRelation.scala

```diff
@@ -49,50 +49,74 @@ private[redshift] case class RedshiftRelation(
     }
   }
 
-  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
-    // Always quote column names:
-    val columns = requiredColumns.map(col => s""""$col"""").mkString(", ")
-    val whereClause = FilterPushdown.buildWhereClause(schema, filters)
-    unloadToTemp(columns, whereClause)
-    makeRdd(pruneSchema(schema, requiredColumns))
-  }
-
   override def insert(data: DataFrame, overwrite: Boolean): Unit = {
     val updatedParams =
       Parameters.mergeParameters(params.parameters updated ("overwrite", overwrite.toString))
     new RedshiftWriter(jdbcWrapper).saveToRedshift(sqlContext, data, updatedParams)
   }
 
-  private def unloadToTemp(columnList: String = "*", whereClause: String = ""): Unit = {
-    val conn = jdbcWrapper.getConnector(params.jdbcDriver, params.jdbcUrl)
-    val unloadSql = unloadStmnt(columnList, whereClause)
-    val statement = conn.prepareStatement(unloadSql)
+  private val tableNameOrSubquery: String = {
+    val unescaped = params.query.map(q => s"($q)").orElse(params.table).get
+    unescaped.replace("'", "\\'")
```

Comment:
Actually, on closer inspection I think that this is actually a mistake when we're handling an empty column query, since in those cases we won't end up adding the extra layer of single quotes. I'm going to add a test to expose this bug and will refactor to fix this.